[AIEX] Add a latency-aware heuristic component to AIEWAWRegRewriter

andcarminati opened the pull request:
This PR brings optimality in terms of pipelined loops for several important kernels.

andcarminati committed:
[AIEX] Add a latency-aware heuristic component to AIEWAWRegRewriter

andcarminati committed:
[AIE2P] Add end-to-end tests related to PostSWP and WAW reg rewrite

andcarminati committed:
[AIE2P] Add base test for AIEWawRegRewriter latency-aware strategy

andcarminati commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -609,6 +631,48 @@ AIEWawRegRewriter::getLastVRegDef(const MachineBasicBlock &MBB) const {
+      if (!Reg.isVirtual())
+        continue;
+
+      auto OperandCycle = ItinData->getOperandCycle(SchedClass, I);
```
Move next to use?

andcarminati commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -609,6 +631,48 @@ AIEWawRegRewriter::getLastVRegDef(const MachineBasicBlock &MBB) const {
+        IsHighLat = true;
+      }
+
+      if (IsHighLat) {
```
Early return?

martien-de-jong replied:
perhaps even `return <closed expression>`

andcarminati committed:
[AIEX] Add temporary heuristic to cutoff some unprofitable cases

Ideally, we should extend with SWP metrics.

andcarminati commented:
With the new cutoff commit:
<img width="4400" height="1800" alt="Core_Insn_Count" src="https://github.com/user-attachments/assets/499aec30-47e6-4978-8aba-e349b38162c7" />

Improvement: 1.11% avg.

krishnamtibrewala commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -609,6 +631,59 @@ AIEWawRegRewriter::getLastVRegDef(const MachineBasicBlock &MBB) const {
+      } else if (MI.mayLoad()) {
+        // If we have an instruction without OperandCycles, it is most probably
+        // a pseudo instruction (no itinerary). In this case, if it is a _split
+        // load, consider it as high latency.
```
check : we should not have multi-slot pseudo here, they have a itinerary with it.

andcarminati replied:
The multislot pseudos are ok here. they use one itinerary as default. The problem that I was facing here was related to the _split variants that are used just during RA, and they don't have itinerary. They also have a different number of operands.

krishnamtibrewala commented:
> With the new cutoff commit: <img alt="Core_Insn_Count" width="2000" height="1800" src="https://private-user-images.githubusercontent.com/40280870/503239111-499aec30-47e6-4978-8aba-e349b38162c7.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjA5ODEzNzIsIm5iZiI6MTc2MDk4MTA3MiwicGF0aCI6Ii80MDI4MDg3MC81MDMyMzkxMTEtNDk5YWVjMzAtNDdlNi00OTc4LThhYmEtZTM0OWIzODE2MmM3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEwMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMDIwVDE3MjQzMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNiYzVhN2U3OWU3YTkyYTUxZTk2ZjdiNTI0MGY5ZDJlN2Y2N2YyMjFkNmUwMzZkOWVjNDk5MTA4MWU4YjQ4Y2ImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.yXu8pyHgy9Ik9S7v4w_VFoOmSERnFuEDPrFExC_1MSs">
> 
> Improvement: 1.11% avg.

Curious to know how was the cut-off decided, and did it lead to reduction in perf, on benchmark where we were getting better performance without the cut-off logic

andcarminati commented:
> > With the new cutoff commit: <img alt="Core_Insn_Count" width="2000" height="1800" src="https://private-user-images.githubusercontent.com/40280870/503239111-499aec30-47e6-4978-8aba-e349b38162c7.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjA5ODEzNzIsIm5iZiI6MTc2MDk4MTA3MiwicGF0aCI6Ii80MDI4MDg3MC81MDMyMzkxMTEtNDk5YWVjMzAtNDdlNi00OTc4LThhYmEtZTM0OWIzODE2MmM3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEwMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMDIwVDE3MjQzMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNiYzVhN2U3OWU3YTkyYTUxZTk2ZjdiNTI0MGY5ZDJlN2Y2N2YyMjFkNmUwMzZkOWVjNDk5MTA4MWU4YjQ4Y2ImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.yXu8pyHgy9Ik9S7v4w_VFoOmSERnFuEDPrFExC_1MSs">
> > Improvement: 1.11% avg.
> 
> Curious to know how was the cut-off decided, and did it lead to reduction in perf, on benchmark where we were getting better performance without the cut-off logic

The original goal of this PR was Conv2D_DW_bf16, but I saw that it also affected Gemm. In this way, I decided to keep those kernels on the heuristic radar. This whole work was based on assembly code analysis (including the cutoff heuristic).

andcarminati commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -361,7 +376,14 @@ bool AIEWawRegRewriter::renameMBBPhysRegs(const MachineBasicBlock *MBB) {
+
+        // Only consider to front physical registers that also
+        // used by high latency operands.
+        if (HighLatencyRegs.count(AssignedPhysReg))
```
If you see here, we reverse the MO ordering by pushing front. This is not by mistake, this helps a lot!

F-Stuckmann replied:
add this as a source code comment

F-Stuckmann replied:
CHECK: We push them to the front so that they get replaced first?

F-Stuckmann replied:
nit: Rewrite the comment to 
`High latency registers should be renamed first, therefore insert them at the front`

F-Stuckmann replied:
Maybe also rewrite for clarity: 

```
const auto InsertPoint = HighLatencyRegs.count(AssignedPhysReg) ? Candidates.begin() : Candidates.end(); 
Candidates.emplace(InsertPoint , &MO, AssignedPhysReg);
```

andcarminati commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -609,6 +631,59 @@ AIEWawRegRewriter::getLastVRegDef(const MachineBasicBlock &MBB) const {
+    }
+  }
+
+  // This metric gives us an idea about the "demand" for high latency registers
```
Ratio regs/instructions.

martien-de-jong commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -361,7 +376,14 @@ bool AIEWawRegRewriter::renameMBBPhysRegs(const MachineBasicBlock *MBB) {
         MCRegister AssignedPhysReg = VRM->getPhys(Reg);
-        Candidates.emplace_back(&MO, AssignedPhysReg);
+
+        // Only consider to front physical registers that also
```
nit: to put in front registers that are also used

F-Stuckmann commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -178,6 +188,9 @@ class AIEWawRegRewriter : public MachineFunctionPass {
   getLastVRegDef(const MachineBasicBlock &MBB) const;
+
+  std::set<MCRegister>
+  getHighOutputLatencyRegs(const MachineBasicBlock *MBB) const;
```
nit: add a comment. 
Is this the total live range of a physical register or on a single Live Interval?

F-Stuckmann replied:
are we looking here at the physical register mapped to a single virtual register def with a high output latency?

andcarminati replied:
One `MCRegister` can also could be mapped to more than one VReg (defs), provided that at least one is high latency.

F-Stuckmann commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -609,6 +631,48 @@ AIEWawRegRewriter::getLastVRegDef(const MachineBasicBlock &MBB) const {
+      bool IsHighLat = false;
+      if (OperandCycle.has_value()) {
+        if (OperandCycle.value() >= MinRegisterLatency)
+          IsHighLat = true;
```
nit: add brackets here to differentiate between the `!OperandCycle.has_value()` case

F-Stuckmann commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -609,6 +631,48 @@ AIEWawRegRewriter::getLastVRegDef(const MachineBasicBlock &MBB) const {
+      if (OperandCycle.has_value()) {
+        if (OperandCycle.value() >= MinRegisterLatency)
+          IsHighLat = true;
+      } else if (MI.mayLoad()) {
```
nit: don't we always have IsHighLat==true when we encounter a load instruction? 
I think we could simplify the conditions to : 

```
auto SetIsHighLat = [&ItinData, &MinRegisterLatency](const MachineInstr &) {
  if (MI.mayLoad())
    return true;

  auto OperandCycle = ItinData->getOperandCycle(SchedClass, I);
  if (!OperandCycle)
    return false;

  return OperandCycle.value() >= MinRegisterLatency;
};
```

andcarminati replied:
Nice suggestion, I will just trust the load itinerary before, because we insert less registers in the set and this could bias the cutoff.

F-Stuckmann commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -639,8 +639,10 @@ std::set<MCRegister> AIEWawRegRewriter::getHighOutputLatencyRegs(
 
   auto *ItinData = MF->getSubtarget().getInstrItineraryData();
   std::set<MCRegister> HighLatRegisters;
+  unsigned NInstrs = 0;
```
nit: why don't we querry `MBB.size()` ?

F-Stuckmann commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -670,6 +672,15 @@ std::set<MCRegister> AIEWawRegRewriter::getHighOutputLatencyRegs(
+  // latency-aware heuristic.
+  // TODO: this should be replaced by more stable metrics related to SWP.
+  if (!HighLatRegisters.empty() &&
+      (((HighLatRegisters.size() * 100 / NInstrs)) < 250 /*calibrated value*/))
```
nit: for readability add a nice name for `(HighLatRegisters.size() * 100 / NInstrs)`. Something like: `HighLatencyRegisterInstrRatio`

martien-de-jong replied:
HighLatencyRegisterCountToInstrCountRatioInPercent

andcarminati replied:
I changed the name according @F-Stuckmann and extended the comment to accommodate @martien-de-jong observation.

F-Stuckmann commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -53,6 +53,16 @@ static cl::opt<bool> PreAlloc(
+                       cl::init(3));
+
+static cl::opt<bool>
+    LatencyAware("aie-realloc-latencyaware", cl::Hidden, cl::init(false),
```
nit: in the last commit set the option to true, since with the cutoff it is now profitable.

andcarminati committed:
[AIEX] Enable AIEWAWRegRewriter latency-aware heuristic

F-Stuckmann commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -609,6 +634,46 @@ AIEWawRegRewriter::getLastVRegDef(const MachineBasicBlock &MBB) const {
+      if (!Reg.isVirtual())
+        continue;
+
+      auto IsHighLatInstrOperand = [&]() {
```
out of curiosity. Why is this lambda not outside of the for loops?

andcarminati replied:
In this case, my goal was to have a `lazy evaluation boolean` where we can state the conditions very explicitly. It is inside so we can capture everything instead of using parameters (clean if statement).

martien-de-jong replied:
I think we can capture most by value.

andcarminati replied:
We use `ItinData` and `MI`, one could be by value and other one by ref. I think I will keep as is to be less verbose.

F-Stuckmann commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -609,6 +634,46 @@ AIEWawRegRewriter::getLastVRegDef(const MachineBasicBlock &MBB) const {
+        return MI.mayLoad();
+      };
+
+      if (IsHighLatInstrOperand()) {
```
nit: reduce the nesting

```
if (!IsHighLatInstrOperand())
     continue;
     
const MCRegister AssignedPhysReg = VRM->getPhys(Reg);
for (MCRegAliasIterator AI(AssignedPhysReg, TRI, true); AI.isValid();
             ++AI)
          HighLatRegisters.insert(*AI);
```

F-Stuckmann commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -671,6 +671,17 @@ std::set<MCRegister> AIEWawRegRewriter::getHighOutputLatencyRegs(
+  // latency-aware heuristic. Basically, we evaluate high latency register count
+  // to instruction count ratio in percent.
+  // TODO: this should be replaced by more stable metrics related to SWP.
+  const int HighLatencyRegisterInstrRatio =
```
could we have a unit case for this?

andcarminati replied:
We have one more test now.

andcarminati committed:
[AIE2P] Add base test for the latency-aware WRWRegRewrite cutoff heuristic

This commit adds an end-to-end test that fails to be scheduled (post-swp) optimally because
of the actual register allocation. This regression is introduced by the latency-aware
WAWRegRewriter strategy.

andcarminati commented on llvm/test/CodeGen/AIE/aie2p/end-to-end/add-att-broadcasting.ll:
```diff
@@ -31,35 +31,36 @@ define dso_local void @add_attribute_bcast(ptr noalias %ifm2, ptr noalias %ifm1,
+; CHECK-NEXT:    vlda.conv.fp32.bf16 cml4, [p0], m1
+; CHECK-NEXT:    vlda.conv.fp32.bf16 cml2, [p1], m0
+; CHECK-NEXT:    vlda.conv.fp32.bf16 cml0, [p1], m0; movx r1, #60
+; CHECK-NEXT:    vlda.conv.fp32.bf16 cml1, [p0], m1; add.nc lc, r0, #-3; vadd.f dm3, dm0, dm1, r1
```
We can reach one more stage now.

andcarminati commented:
Complete QoR view:

<img width="4400" height="1800" alt="Core_Insn_Count" src="https://github.com/user-attachments/assets/14bf3cb7-5167-476a-84db-95aa15a6157d" />

<img width="4400" height="1800" alt="Core_StackSize_absolute" src="https://github.com/user-attachments/assets/2536e9d7-d01d-4bfb-b00d-3fc0b0aa115f" />

<img width="4400" height="1800" alt="Core_PMSize_absolute" src="https://github.com/user-attachments/assets/d6acd9a5-027d-4fb1-b89a-92904ab30bdb" />

(PM increase when we increase the number of SWP stages)

F-Stuckmann commented:
What happens with stack and PM ?

andcarminati commented:
> What happens with stack and PM ?

I added to the previous comment.

F-Stuckmann reviewed (commented):
LGTM
